File: backend/services/payment-service/internal/model/payment.go
package model

import (
	"encoding/json"
	"time"
)

// Payment 对应 payments 表：一次「向用户收这笔钱」的尝试。
//
// 一个订单可以有多个支付单：第一次支付失败、超时被关，用户可以再发起一次，那是新的一行。
// 但**成功只能有一张**（payments_one_succeeded_per_order 部分唯一索引），因为订单只能被
// 收一次钱。所以「这次能不能建单」不是看有没有既存支付单，而是看有没有成功的那张。
type Payment struct {
	ID        string `db:"id"`
	PaymentNo string `db:"payment_no"`
	// 老库（MongoDB ObjectID 的 24 位 hex），仅迁移过来的行有值。
	LegacyID *string `db:"legacy_id"`
	// 订单号（orders.order_no），值引用：订单事实的归属方是订单服务，这里只存一个值。
	OrderNo string `db:"order_no"`
	UserID  string `db:"user_id"`
	// 应付总额，单位为分；等于各笔出资之和。
	Amount int64 `db:"amount"`
	// 走哪条渠道，值是渠道在**代码里**的名字（catalog 里的 Channel.Provider，今天只有一个
	// `ums`）；纯账户出资（咖啡豆）没有第三方，这里是空串。
	//
	// 它从前是一个指向 payment_channels 的外键。那张表没了之后，这一列仍然值得留：后台
	// 列表要显示「这笔走的哪条渠道」，对账要按渠道分组，而这两件事都不该要求再查一次表。
	Provider string `db:"provider"`
	// 用户在收银台上选的那一个支付方式的 **code**（catalog 里的常量，如 `ums_h5_wechat`）。
	//
	// 与 Provider 的关系是「一个渠道下挂着几条方式」：小程序那条与 H5 那四条同属 ums，
	// 咖啡豆那条不属于任何渠道。所以这两列不是一回事，缺任一列都答不出「用户当初选了什么」。
	PaymentMethod string `db:"payment_method"`
	// created / pending / succeeded / failed / closed / expired，见下面的常量。
	Status string `db:"status"`
	// 渠道收银台与账单上显示的商品描述。
	Subject string `db:"subject"`
	// 渠道附加数据（小程序 openid、设备号等），**不放密钥**。
	Attach json.RawMessage `db:"attach"`
	// 渠道侧的交易号，回调与查单都以它为准。
	ProviderTransactionID string `db:"provider_transaction_id"`
	FailureCode           string `db:"failure_code"`
	FailureMessage        string `db:"failure_message"`
	// 发起方给的幂等号，落到这一行上供排查；真正挡重的是 payment_idempotency_keys。
	RequestID string     `db:"request_id"`
	ExpiresAt *time.Time `db:"expires_at"`
	// 账户出资（纯豆）扣豆的账变 ID，在**扣豆返回的那一刻**就落库、先于结算。
	//
	// 非空 = 这笔钱确实动过。它有两个用途：超时关单不碰这样的单（把钱已经动过的单当成
	// 没付过一样关掉、把出资标成 released，是一次记账上的撒谎），补偿任务把它们结算掉。
	// 渠道支付永远是 nil——这条路上没有账户域可扣。见 payments.account_entry_id 的列注释。
	AccountEntryID *string `db:"account_entry_id"`
	// 扣豆成交的时刻（账户域那笔账变的发生时刻）。账户出资没有第三方给成交时间，
	// 结算与补偿都用它，而不是拿处理请求的那一刻顶上。
	AccountFundedAt *time.Time `db:"account_funded_at"`
	// 渠道给的成交时间，不是我们记账的时间。
	PaidAt    *time.Time `db:"paid_at"`
	ClosedAt  *time.Time `db:"closed_at"`
	CreatedAt time.Time  `db:"created_at"`
	UpdatedAt time.Time  `db:"updated_at"`
}

// 支付单状态，与 payments.status 的 CHECK 逐字一致。
const (
	// PaymentCreated：已建单、还没向渠道发起。
	PaymentCreated = "created"
	// PaymentPending：已向渠道发起（或已返回支付参数），等回调或轮询。
	PaymentPending = "pending"
	// PaymentSucceeded / PaymentFailed 是终态。
	PaymentSucceeded = "succeeded"
	PaymentFailed    = "failed"
	// PaymentClosed：订单取消等主动关单；PaymentExpired：待支付超时。
	PaymentClosed  = "closed"
	PaymentExpired = "expired"
)

// IsPaymentStatus 判断一个值是不是合法的支付单状态。
//
// 后台列表的 status 筛选用它**在进 SQL 之前**挡下打错的值（见 controller 的 listPayments）：
// 一个不在词表里的值送进 SQL 不报错，只会安静地返回空列表——而运营会把它当成「这段时间
// 真的没有成功的单」拿去对账。
//
// 写成 switch 而不是从上面那六个常量拼一个切片：这里要的是「整份摆出来」，多一个常量时
// 少写一行会当场编译不过（重复的 case），而漏在一个切片里不会。
func IsPaymentStatus(status string) bool {
	switch status {
	case PaymentCreated, PaymentPending, PaymentSucceeded, PaymentFailed, PaymentClosed, PaymentExpired:
		return true
	default:
		return false
	}
}

// 出资类型那套词表（wechat / unionpay / coffee_bean / wallet / other）**已经退场**，
// 连同 payments.funding_type 那一列一起。
//
// 从前它是「这笔钱从哪个通道出」的归纳，与「用户点了哪个支付方式」分两处记。代价是加一种
// 支付方式要同时在 catalog 里给 code、再在这套词表里给它找一档——找不到就得改两个库的 DDL，
// 支付宝落 `other` 就是这么来的（后台订单列表把一笔支付宝单显示成「其他」）。
//
// 今天只剩一套值：catalog 的 code。出资行、记账流水、退款出资行都存它。要判「这笔钱走不走
// 渠道」，看的是这条方式有没有渠道（catalog.Method.ChannelCode / 路由上的 Channel），
// 不是看某个词表值——唯一的例外见 service/refund.go 里那段关于 coffee_bean 的说明。

File: admin-web/src/services/payment.ts
import { request } from '@umijs/max';
import type { PageQuery, PageResult } from './pagination';

/**
 * 支付域的接口。字段名与后端 dto（payment-service internal/dto/response.go 的 json tag）
 * 逐字对应，同时也是页面上的 dataIndex——改一处要同批改三处，否则 TypeScript 不报错，
 * 只是整列空白。
 *
 * 金额一律是「分」的整数（int64 到了 JS 里就是 number）。这一层不换算，页面按元展示。
 *
 * **这个文件里全是 GET。** 支付单是钱的既成事实，后面挂着出资行、记账流水、渠道调用、
 * 回调通知四张只增表——关单、改状态、重放回调、发起退款都要先有退款与对账的语义，那几件事
 * 本轮没做，后端那几条路径连路由都没注册，写出来只会得到一个 404。
 *
 * 收钱那一侧**在这个文件里已经没有东西了**：支付方式与渠道不再是库里的两行可写配置，而是
 * payment-service internal/catalog 里的三个常量（咖啡豆 / 银联商务小程序 / 银联商务 H5）。
 * 第四种「取货码」不在那里：它扣的是这台设备的咖啡余额，由 partner-service 验签后转
 * order-service 直接落单，压根不经过支付服务——所以支付域这边连它的名字都没有。
 * 运营没有可配的东西，后台也就没有可写的表单——原来那一页
 * （/payments/methods）连同它那几个写接口一起删了。
 */

// ——— 枚举 ———
// 取值来自 payment-service 的 model 常量（各表的 CHECK 约束），不是接口给的——接口回的
// 就是库里那些英文码。所以**改枚举必须同时改迁移和这里**，加了新码而这里没登记，界面上
// 就退回显示原始码。文案表在 services/paymentLabels.ts。

/** payments.status：支付单走到哪一步了 */
export type PaymentStatus =
  | 'created'
  | 'pending'
  | 'succeeded'
  | 'failed'
  | 'closed'
  | 'expired';

/** payment_fundings.status：一条出资行走到哪一步 */
export type PaymentFundingStatus = 'reserved' | 'succeeded' | 'failed' | 'released' | 'reversed';

/**
 * payment_methods.action：这一档方式怎么把钱收上来。
 *
 * 前端**认的是 action 不是 code**——加一档新方式（换个 code）只要 action 是认得的六个
 * 之一就不用改代码。所以这一列在配置页上必须显眼，它是这张表唯一的可执行语义。
 */
export type PaymentMethodAction =
  | 'jump_miniapp'
  | 'native_pay'
  | 'direct_pay'
  | 'qrcode'
  | 'h5'
  | 'account';

export type PaymentMethodStatus = 'enabled' | 'disabled';

/**
 * payment_channels.mode：沙箱还是真实渠道。
 *
 * 只有两个值，但它是这一页最要紧的一列：`live` 的渠道行连着真钱。
 */
export type PaymentChannelMode = 'sandbox' | 'live';

/** payment_channels.status：legacy_readonly = 老系统留下的只读渠道，能查不能新用 */
export type PaymentChannelStatus = 'enabled' | 'disabled' | 'legacy_readonly';

/** payment_state_transitions.actor_type：是谁做的这次迁移 */
export type PaymentActorType = 'user' | 'merchant' | 'admin' | 'system';

/** payment_state_transitions.aggregate_type：这条流水记的是谁的迁移 */
export type PaymentAggregateType =
  | 'payment'
  | 'funding'
  | 'refund'
  | 'agreement'
  | 'charge'
  | 'reconciliation';

/** payment_transactions.kind：这笔账是收、退还是冲正 */
export type PaymentTransactionKind = 'payment' | 'refund' | 'reversal';

/** payment_transactions.direction：钱进还是钱出 */
export type PaymentTransactionDirection = 'in' | 'out';

/** payment_provider_calls.operation：这次出网是在调渠道的哪件事 */
export type ProviderOperation =
  | 'create'
  | 'query'
  | 'close'
  | 'refund'
  | 'query_refund'
  | 'agreement_sign'
  | 'agreement_charge'
  | 'agreement_terminate'
  | 'reconcile';

/** payment_provider_calls.result：这次出网的结果。unknown = 没拿到明确答复 */
export type ProviderResult = 'success' | 'failed' | 'timeout' | 'unknown';

/** payment_notifications.status：一条回调通知处理到哪一步了 */
export type NotificationStatus = 'received' | 'processed' | 'ignored' | 'failed';

/**
 * 库里是 jsonb、后端原样透传的列，形状随业务变，所以是 unknown 而不是具体结构。
 *
 * 页面**不铺成一列一列**，原样 `<pre>` 贴出来：它们是排查时核对用的原始凭证
 * （渠道到底回了什么、这一档方式的启动参数是什么），不是给人扫的表格。
 */
type RawJSON = unknown;

// ——— 支付单 ———

/**
 * 一张支付单。**列表与详情是同一份形状**（后端也只有一个 dto），所以这里的类型两者共用，
 * 列定义抄过来也不会错位。
 */
export type Payment = {
  id: string;
  paymentNo: string;
  /** 老系统迁移过来的单号，没迁移过的是 null（本仓库自己的单都是 null）。 */
  legacyId: string | null;
  orderNo: string;
  userId: string;
  /** 分。 */
  amount: number;
  status: PaymentStatus;
  subject: string;
  /** 渠道附加数据（小程序 openid、设备号等），不放密钥。 */
  attach: RawJSON;
  providerTransactionId: string;
  /**
   * 失败码与失败话术。成功单上是**空串**而不是 null，而 closedAt 那类时间是**真 null**，
   * 两种都当空显示（页面上都是 `—`）。
   */
  failureCode: string;
  failureMessage: string;
  requestId: string;
  /**
   * 渠道与方式的 code 与中文名。两个 code 就在支付单自己那两列上（provider /
   * payment_method），名字由服务层照代码里的目录补上——**没有 uuid**：渠道与方式已经不是
   * 两张表了，没有什么可 JOIN 的。账户出资那条路上渠道是空串，于是 channelCode /
   * channelName 都是空，页面显示 `—`。
   */
  channelCode: string;
  channelName: string;
  methodCode: string;
  methodName: string;
  /** 非空 = 这笔钱确实动过账户域（纯豆支付扣豆的那笔账变）。 */
  accountEntryId: string | null;
  accountFundedAt: string | null;
  expiresAt: string | null;
  paidAt: string | null;
  closedAt: string | null;
  createdAt: string;
  updatedAt: string;
};

/**
 * 一条出资行：这笔钱按来源拆成的一行。
 *
 * lineType 就是**这笔出资的支付方式 code**，与同一张支付单上的 methodCode 是同一个值
 * （出资渠道那套词表已经退场，见 migrations/payment）。所以它是个自由字符串：合法的
 * 取值就是 catalog 里那几条，前端没有一份能对齐的枚举，中文名走 paymentMethodLabels。
 */
export type PaymentFunding = {
  id: string;
  lineNo: number;
  lineType: string;
  amount: number;
  status: PaymentFundingStatus;
  providerTransactionId: string;
  failureCode: string;
  accountEntryId: string | null;
  succeededAt: string | null;
  reversedAt: string | null;
  createdAt: string;
  updatedAt: string;
};

/**
 * 一条记账流水。这张表**只增不改**：一笔出资成功写一条 in，将来退款写一条 out，冲正再写
 * 一条。所以同一条出资行会有多行流水，靠 kind 与 direction 区分——页面上不合并。
 */
export type PaymentTransaction = {
  id: string;
  legacyId: string | null;
  kind: PaymentTransactionKind;
  paymentNo: string;
  refundNo: string;
  fundingLineNo: number;
  /** 这条流水对应的支付方式 code，与出资行 lineType 同一套值。见上面 PaymentFunding 的说明。 */
  lineType: string;
  direction: PaymentTransactionDirection;
  amount: number;
  /** 走的是哪家（ums / 空）。这一列是**文本**：渠道表已经删了，没有可指的 id。 */
  provider: string;
  providerTransactionId: string;
  accountEntryId: string | null;
  occurredAt: string;
  createdAt: string;
};

/**
 * 一条状态流转。
 *
 * `reason` 是**中英混着的诊断串**（`payment created` / `provider reported payment failed:
 * 用户取消` / `咖啡豆余额不足，需要 3400 分`），页面原样显示，不做码表映射——那种映射一定
 * 映射错，而它承载的恰恰是「当时到底发生了什么」这条唯一的一手信息。
 */
export type PaymentTransition = {
  id: string;
  aggregateType: PaymentAggregateType;
  aggregateId: string;
  fromStatus: string;
  toStatus: string;
  reason: string;
  requestId: string;
  actorType: PaymentActorType;
  actorId: string | null;
  metadata: RawJSON;
  createdAt: string;
};

/** 一次出网调用。请求/响应摘要原样显示（已由 provider 层做过脱敏，这里不再二次裁剪）。 */
export type PaymentProviderCall = {
  id: string;
  provider: string;
  operation: ProviderOperation;
  paymentNo: string;
  refundNo: string;
  agreementNo: string;
  requestId: string;
  traceId: string;
  /** 同一件事的第几次尝试。 */
  attemptNo: number;
  requestSummary: RawJSON;
  responseSummary: RawJSON;
  httpStatus: number | null;
  providerCode: string;
  providerMessage: string;
  result: ProviderResult;
  durationMs: number | null;
  createdAt: string;
};

/**
 * 一条回调通知。
 *
 * `body` 是渠道发来的原始报文。后端已经把它从 bytea 显式转成 string（不转的话 JSON 出的是
 * base64，页面上是乱码中的乱码），**非 UTF-8 的字节会被换成 U+FFFD**，所以页面上偶尔会
 * 看到替换字符——那不是解析错了。
 */
export type PaymentNotification = {
  id: string;
  provider: string;
  notificationId: string;
  eventType: string;
  paymentNo: string;
  refundNo: string;
  body: string;
  bodySha256: string;
  headers: RawJSON;
  signatureVerified: boolean;
  status: NotificationStatus;
  failureReason: string;
  receivedAt: string;
  processedAt: string | null;
};

/**
 * 支付单详情：一次把六个页签要的都给全。
 *
 * 六块在同一个只读事务里读出来（后端 repository.GetPaymentDetail），所以它们之间不会出现
 * 「支付单 succeeded 而出资行还停在 reserved」这种从未存在过的中间态。
 */
export type PaymentDetail = {
  payment: Payment;
  fundings: PaymentFunding[];
  transactions: PaymentTransaction[];
  transitions: PaymentTransition[];
  providerCalls: PaymentProviderCall[];
  notifications: PaymentNotification[];
};

// ——— 查询参数 ———

/**
 * 支付单筛选。
 *
 * paymentNo / orderNo 后端走 `ILIKE '%…%'`（**模糊**，与订单域的等值不同）：运维手里多半
 * 只有单号的一截。userId 是 uuid 列上的**等值**比较，且后端会先校验是不是合法 uuid——
 * 不是的话回 400 而不是让 Postgres 报类型错误。
 *
 * status / methodCode 走枚举白名单：打错一个字母回 400，不会静默返回空列表
 * （那会与「这段时间真的一张成功的单都没有」长得一模一样）。
 */
export type PaymentQuery = PageQuery & {
  paymentNo?: string;
  orderNo?: string;
  userId?: string;
  status?: PaymentStatus;
  /**
   * 按支付方式筛（`payments.payment_method`）。值域是 catalog 的 code——服务端拿那张目录
   * 校验，不是这里写死的联合类型：加一种支付方式只加 catalog 一条，别在这里补一次枚举。
   */
  methodCode?: string;
  /** 创建时间的闭开区间 [createdFrom, createdTo)，必须是带时区的 RFC3339（用 toRFC3339 转）。 */
  createdFrom?: string;
  createdTo?: string;
};

// ——— 接口 ———

export async function listPayments(params?: PaymentQuery) {
  return request<PageResult<Payment>>('/api/v1/admin/payments', { params });
}

/**
 * 支付单详情。路径参数是**支付单号**不是 uuid。
 *
 * 单号查无此单时后端回 404（`paymentNo` 是文本列，不像订单详情那样会因非 uuid 回 500），
 * 所以详情页拿到的是路由参数直接查，不必先校验形状。
 */
export async function getPaymentDetail(paymentNo: string) {
  return request<PaymentDetail>(`/api/v1/admin/payments/${paymentNo}`);
}
